DAOS-15204 cart: add crt_perf utils

soumagne opened the pull request:
### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/1/execution/node/376/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/1/execution/node/383/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/1/execution/node/305/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/1/execution/node/361/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/1/execution/node/349/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/1/execution/node/327/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/2/execution/node/295/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/2/execution/node/380/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/2/execution/node/278/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/2/execution/node/370/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/2/execution/node/383/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/2/execution/node/324/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/3/execution/node/287/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/3/execution/node/310/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/3/execution/node/315/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/3/execution/node/344/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/4/execution/node/383/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/4/execution/node/378/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/4/execution/node/322/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/5/execution/node/363/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/5/execution/node/278/log

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1532 @@
+	rc = crt_init_opt(listen ? CRT_PERF_GROUP_ID : NULL, crt_init_flags, &crt_init_options);
+	CRT_PERF_CHECK_D_ERROR(error, rc, "could not init CART");
+
+	if (attach_info_path) {
```
from past experience might be helpful to debug/info print the attach info path

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1532 @@
+	CRT_PERF_CHECK_ERROR(cb_info->cci_rc != 0, out, info->rc, cb_info->cci_rc,
+			     "callback failed");
+
+	if (info->cb) {
```
surprised checkpatch doesnt complain about single statement being wrapped in  { } 's

daosbuild1 commented:
Test stage Unit Test on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13825/7/testReport/

daosbuild1 commented:
Test stage Unit Test with memcheck on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13825/7/testReport/

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/11/execution/node/816/log

frostedcmos commented on src/cart/crt_internal_types.h:
```diff
@@ -150,6 +150,8 @@ struct crt_gdata {
 	unsigned int             cg_provider_is_primary : 1;
 	/** use legacy progress method */
 	bool                     cg_progress_legacy;
+	/** use single thread to access context */
```
might want to add new field to be dumped in crt_gdata_dump(void). even though we dump envs automatically, someone might set new field via init_opt, and would be good to see that in logs.

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13825/12/execution/node/804/log

daosbuild3 commented:
Test stage NLT completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-13825/5/execution/node/673/log

soumagne committed:
DAOS-15204 cart: add crt_perf utils

Add crt_rate, crt_bw_read/write, crt_perf_server executables

CRT perf utils can be run with and without MPI support (auto-detected)

Add ability to run crt_rate against DAOS server

Fix some missing static qualifiers in crt_rpc.c

Signed-off-by: Jerome Soumagne <jerome.soumagne@hpe.com>

soumagne committed:
DAOS-15204 build: add crt_perf to packaging

Signed-off-by: Jerome Soumagne <jerome.soumagne@hpe.com>

daosbuild3 commented:
Test stage NLT completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-13825/7/testReport/

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-13825/7/execution/node/1215/log

frostedcmos commented on src/utils/crt_perf/crt_bw_read.c:
```diff
@@ -0,0 +1,74 @@
+
+	info = &perf_info.context_info[0];
+
+	/* Set RPC requests */
```
Not quite sure what this means, maybe elaborate a bit?

frostedcmos commented on src/utils/crt_perf/crt_bw_read.c:
```diff
@@ -0,0 +1,74 @@
+	for (size = MAX(1, perf_info.opts.buf_size_min); size <= perf_info.opts.buf_size_max;
+	     size *= 2) {
+		rc = crt_perf_run_bw(&perf_info, info, size,
+				     (size > CRT_PERF_LARGE_SIZE) ? CRT_PERF_SKIP_LARGE
+								  : CRT_PERF_SKIP_SMALL,
```
A comment here would be nice explaining what this skipping is

frostedcmos commented on src/utils/crt_perf/crt_perf.h:
```diff
@@ -0,0 +1,117 @@
+	struct crt_perf_mpi_info      mpi_info;
+	crt_group_t                  *ep_group;
+	uint32_t                      ep_ranks;
+	uint32_t                      ep_tags;
```
is this num_ranks/num_tags?

frostedcmos commented on src/utils/crt_perf/crt_perf.h:
```diff
@@ -0,0 +1,117 @@
+/*****************/
+/* Public Macros */
+/*****************/
+
```
a comment here would help to explain what this skipping is

frostedcmos commented on src/utils/crt_perf/crt_perf_server.c:
```diff
@@ -0,0 +1,166 @@
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
```
is this for sched.h?

frostedcmos commented on src/utils/crt_perf/crt_perf_server.c:
```diff
@@ -0,0 +1,166 @@
+		size_t i;
+
+		progress_threads =
+		    (pthread_t *)malloc(sizeof(*progress_threads) * info.opts.context_max);
```
you could use D_ALLOC_ARRAY here :)

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+
+struct crt_perf_request {
+	int32_t expected_count;         /* Expected count */
+	int32_t complete_count;         /* Completed count */
```
should this be 'completed_count' for consistency?

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+	int32_t expected_count;         /* Expected count */
+	int32_t complete_count;         /* Completed count */
+	int     rc;                     /* Callback return code */
+	bool    done;                   /* Request */
```
request?

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+};
+
+static struct crt_proto_rpc_format crt_perf_rpcs[] = {
+    {
```
seeing this repeated many times wants me to turn this part into a RPC_FMT(fmt, cb) or some such macro:)

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+	}
+
+	if ((argc - optind) > 1) {
+		crt_perf_usage(argv[0]);
```
missing crt_perf_free_options() here for any options strdup-ed up to this point

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+	if (opts->comm != NULL) {
+		free(opts->comm);
+		opts->comm = NULL;
+	}
```
or you could use D_STRDUP/D_FREE where D_FREE does this for you

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+		CRT_PERF_CHECK_ERROR(info->ep_tags == 0, error, rc, -DER_INVAL,
+				     "ep tags cannot be zero");
+
+		printf("# %" PRIu32 " target rank(s) read - %" PRIu32 " tag(s)\n", info->ep_ranks,
```
you might want to change all printfs to some other macro that would also log, as stdout is lost in some tests in ci... in crt_utils we use DBG_PRINT for that

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+		size_t  alloc_size = info->buf_size_max * info->buf_count;
+		d_iov_t iov = {.iov_buf = NULL, .iov_buf_len = alloc_size, .iov_len = alloc_size};
+		d_sg_list_t sgl = {.sg_nr = 1, .sg_nr_out = 0, .sg_iovs = &iov};
+
```
would be nice to have a verbose option with info like alloc size printed here:)

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+		d_sg_list_t sgl = {.sg_nr = 1, .sg_nr_out = 0, .sg_iovs = &iov};
+
+		/* Prepare buf */
+		info->bulk_bufs[i] = aligned_alloc(page_size, alloc_size);
```
should we also have an option for non-aligned allocs using generic malloc? real world user buffers arent always page aligned, so perf testing with ideal buffers might not enough

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+}
+
+static int
+crt_perf_proc_iovec(crt_proc_t proc, void *data)
```
as discussed this might not be needed

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+}
+
+static int
+crt_perf_proc_tags(crt_proc_t proc, void *data)
```
it's non-intuitive to me what you are proc-ing here. what are tags?

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+			free(info->bulk_bufs[i]);
+		free(info->bulk_bufs);
+		info->bulk_bufs = NULL;
+	}
```
why not call crt_perf_bulk_free() from below?

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+{
+	size_t i;
+
+	if (info->local_bulk_handles != NULL) {
+		for (i = 0; i < info->handle_max; i++)
+			(void)crt_bulk_free(info->local_bulk_handles[i]);
+		free(info->local_bulk_handles);
+		info->local_bulk_handles = NULL;
+	}
+
+	if (info->bulk_bufs != NULL) {
+		for (i = 0; i < info->handle_max; i++)
+			free(info->bulk_bufs[i]);
+		free(info->bulk_bufs);
+		info->bulk_bufs = NULL;
+	}
```
part of me wants to combine this into a single loop:)

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+	if (info->rpc_buf != NULL) {
+		free(info->rpc_buf);
+		info->rpc_buf = NULL;
+	}
```
D_FREE here would make it much cleaner imo:)

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+	return;
+}
+
+static void
+crt_perf_bulk_push_cb(crt_rpc_t *rpc)
+{
+	crt_perf_bulk_common(rpc, CRT_BULK_PUT);
+}
+
+static void
+crt_perf_bulk_pull_cb(crt_rpc_t *rpc)
+{
+	crt_perf_bulk_common(rpc, CRT_BULK_GET);
+}
```
not sure what wrapping those adds here, might as well call crt_perf_bulk(rpc, <type>) inline imo

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+			for (i = 0; i < info->buf_count; i++) {
+				char *buf_p =
+				    (char *)sgl.sg_iovs[0].iov_buf + info->buf_size_max * i;
+				rc = crt_perf_verify_data(buf_p,
+							  bulk_cb_info->bci_bulk_desc->bd_len);
```
this is rather strange formatting

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+}
+
+void
+crt_perf_rpc_set_req(const struct crt_perf_info *perf_info, struct crt_perf_context_info *info)
```
this feels like needs a better naming, maybe like rpc_set/init_endpoints ?

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,1882 @@
+	int                         rc;
+
+	/* Prepare buf */
+	info->rpc_buf = aligned_alloc(page_size, opts->buf_size_max);
```
same as in the other place -- would be good to have non-ideal-aligned-alloc option

frostedcmos commented on src/utils/crt_perf/crt_perf_mpi.c:
```diff
@@ -0,0 +1,412 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
```
Being part of cart/daos -- feels like this should be using dpar.h instead of having its own mpi layer

frostedcmos reviewed (commented):
mostly minor inline comments.
as discussed offline -- as part of this PR would need to add a simple sanity test that uses crt_perf_server/client in CI

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-13825/8/execution/node/984/log

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-13825/9/testReport/

daosbuild3 commented:
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-13825/9/execution/node/1430/log

frostedcmos commented on src/cart/crt_rpc.c:
```diff
@@ -113,6 +113,48 @@ crt_hdlr_ctl_fi_attr_set(crt_rpc_t *rpc_req)
+
+	rc = crt_reply_send(rpc_req);
+	if (rc != 0)
+		D_ERROR("crt_reply_send() failed. rc: %d\n", rc);
```
We should get into habit of using DL_* variant of macros when rc is present, like DL_ERROR() here
similar comment for few other places where D_ERROR is used

soumagne replied:
I don't disagree but you want me to fix the entire file ? 😅 I was copying what was already there to be consistent. If you ask me, I think we should start using something like:
```C
/* Check for D_ ret value and goto label */
#define CRT_CHECK_D_ERROR(label, rc, ...)                                                     \
	do {                                                                                       \
		if (unlikely(rc != 0)) {                                                           \
			DL_ERROR(rc, __VA_ARGS__);                                                 \
			goto label;                                                                \
		}                                                                                  \
	} while (0)
```

frostedcmos commented on src/cart/crt_rpc.c:
```diff
@@ -113,6 +113,48 @@ crt_hdlr_ctl_fi_attr_set(crt_rpc_t *rpc_req)
+	const int *buf_ptr = (const int *)buf;
+	size_t     i;
+
+	/* Skip first integer (used for checking rank) */
```
feels like a hack a bit:) why not just add rank separately in input payload?

soumagne replied:
yeah ignore that for now, you should keep in mind that I'm also trying to align these benchmarks with the HG ones to keep the code structure relatively similar and make it easier to backport changes between both. I may get rid of that in the cart ones since cart already has rank checking though we could keep it also for sanity checks.

frostedcmos commented on src/cart/crt_rpc.c:
```diff
@@ -113,6 +113,48 @@ crt_hdlr_ctl_fi_attr_set(crt_rpc_t *rpc_req)
+{
+	const int *buf_ptr = (const int *)buf;
+	size_t     i;
+
```
probably want to assert on  buf != null, idx_p etc.. to make coverity happy

soumagne replied:
ok I'll have a look

frostedcmos commented on src/cart/crt_rpc.h:
```diff
@@ -612,6 +622,12 @@ CRT_RPC_DECLARE(crt_ctl_log_set, CRT_ISEQ_CTL_LOG_SET, CRT_OSEQ_CTL_LOG_SET)
 CRT_RPC_DECLARE(crt_ctl_log_add_msg, CRT_ISEQ_CTL_LOG_ADD_MSG,
 		CRT_OSEQ_CTL_LOG_ADD_MSG)
 
+#define CRT_ISEQ_PERF_RATE /* input fields */ ((d_iov_t)(iov)CRT_VAR)((uint8_t)(verify)CRT_VAR)
```
there is a method with comments of /* clang on/off */ to disable clang-format, we should do it for this file, so that we dont end up with changed lines

soumagne replied:
yeah we probably should

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,2113 @@
+#define CRT_PERF_GROUP_ID      "crt_perf"
+
+/* matches crt_rpc.h */
+#define CRT_OPC_PERF_BASE      0xF5000000UL
+#define CRT_PROTO_PERF_VERSION 1
+#define CRT_PERF_DAOS_ID(x)    CRT_PROTO_OPC(CRT_OPC_PERF_BASE, CRT_PROTO_PERF_VERSION, x)
+
+#define CRT_PERF_BASE_OPC      0x010000000
+#define CRT_PERF_RPC_VERSION   1
+#define CRT_PERF_ID(x)         CRT_PROTO_OPC(CRT_PERF_BASE_OPC, CRT_PERF_RPC_VERSION, x)
```
would be nice not have to rely on matching those with crt_rpc.h, perhaps look into what self-test did to avoid that

soumagne replied:
yeah I'm just not sure I like the way it includes internal stuff, I'll think about it.

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,2113 @@
+
+struct crt_perf_bulk_init_info {
+	crt_bulk_t bulk;
+	uint32_t   bulk_op;
```
few structs missing comments/consistency

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,2113 @@
+	while ((opt = getopt_long(argc, argv, crt_perf_short_options, crt_perf_long_options,
+				  NULL)) != -1) {
+		switch (opt) {
+		case 'c': /* comm */
+			opts->comm = strdup(optarg);
```
i read this in few places now, and i still dont know what 'comm' is:)

soumagne replied:
resolving since you said you know :)

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,2113 @@
+	printf("usage: %s [OPTIONS]\n", execname);
+	printf("    OPTIONS\n");
+	printf("    -h, --help           Print a usage message and exit\n");
+	printf("    -c, --comm           Select transport plugin\n"
+	       "                         Available plugins: ofi, ucx, etc\n");
```
now i get what comm is, but why make it separate option instead of following cart style and specifying provider as:  [comm]+[provider] ?

soumagne replied:
this is mostly to match the HG benchmarks, I actually made it so that it matches both formats so you can pass either and do -p or -c ofi+tcp or pass it separately using -c ofi and -p tcp options.

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,2113 @@
+		size_t i;
+		for (i = 0; i < info->handle_max; i++)
+			crt_bulk_free(info->remote_bulk_handles[i]);
+		free(info->remote_bulk_handles);
```
this was allocated with D_ALLOC_ARRAY, should use D_FREE to free

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,2113 @@
+	printf("    -M, --mbps           Output in MB/s instead of MiB/s\n");
+	printf("    -f, --hostfile       Specify attach info path\n");
+	printf("    -g, --group_id       Specify group ID (default:%s)\n", CRT_PERF_GROUP_ID);
+	printf("    -a, --daos_agent     Use DAOS agent (experimental)\n");
```
what is experimental about using agent?

soumagne replied:
meaning it's not production ready and only crt_rate works with it.

frostedcmos commented on src/utils/crt_perf/crt_perf.c:
```diff
@@ -0,0 +1,2113 @@
+	uint32_t ep_rank, ep_tag;
+	int      rc;
+
+	if (perf_info->opts.daos_agent) {
+		D_INFO("Skipping done RPC (DAOS agent)\n");
+		return 0;
+	}
```
not quite sure why agent-enabled option skips this part, perhaps add a comment here

soumagne replied:
because when using the agent we attach to an existing DAOS server so we don't send an RPC to finalize the server in this case.

frostedcmos reviewed (commented):
minor feedback inline

daosbuild3 commented:
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-13825/12/execution/node/1751/log
